### ISSUE TITLE
Generic instance parameter becomes unknown when passed to `type`

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#1210]
**Describe the bug**
The parameter of an instance of a generic class becomes unknown when attempting to derive the class from the instance by calling `type`.  This also happens when the parameter is concrete.

**To Reproduce**
Type check the snippet below.

**Expected behavior**
That the type of `qux` would be `type[Foo[T]]` and that no error would be generated.

**Screenshots or Code**
```python
# pyright: strict

from typing import Generic, Type, TypeVar

T = TypeVar('T')

class Foo(Generic[T]):
    def __init__(self, value: T) -> None:
        ...

def bar(value: T) -> Type[Foo[T]]:
    baz = Foo(value)
    # Type of "qux" is partially unknown
    #   Type of "qux" is "Type[Foo[Unknown]]"
    qux = type(baz)
    # Return type, "Type[Foo[Unknown]]", is partially unknown
    return qux

```

**VS Code extension or command-line**
Either, 1.1.89

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is now addressed in pyright 1.1.90, which I just published. It will also be included in the next release of pylance.

**Comment by layday** [comment]
Thank you, that fixed the issue with invoking `type` directly but it still can't be used as a callback it seems:

```py
# pyright: strict

from typing import Callable, Type, TypeVar

T = TypeVar('T')
U = TypeVar('U')


def foo(value: T, fn: Callable[[T], U]) -> U:
    ...

reveal_type(type(''))  # Ok

bar = foo('', type)
reveal_type(bar)  # Type of "bar" is "Unknown"
```

And because it's unknown, it can't be specialised either:

```py
baz: Type[str] = bar
reveal_type(baz)  # Type of "baz" is "Unknown"
```

One would have to resort to `cast`ing.

**Comment by erictraut** [comment]
Yeah, that's correct. Handling of `type` necessarily involves a bunch of special-case logic, so it can't be treated like class with a callable constructor in this context. It's not even defined as generic in the builtins type stub.
